File: tilelang/tileview/tileview.py
from __future__ import annotations
import tvm_ffi
from tvm.ir import Node
from tvm.tir import Buffer, BufferLoad, BufferRegion
from tilelang import _ffi_api
from tilelang.layout.swizzle import _get_buffer_info


@tvm_ffi.register_object("tl.TileView")
class TileView(Node):
    """
    TileView represents how a buffer is partitioned into tiles for processing.

    A TileView captures:
    - buffer_shape: The original shape of the buffer
    - tile_shape: The shape of each tile (e.g., (16, 32) for a 2D tile)
    - index_map: Which dimensions of the buffer are tiled
    - tiled_buffer_shape: The resulting shape after tiling

    For Sunmmio target, the Tile unit processes fixed-size 2D tiles with
    constraints: width must be 32, height can be 8/16/32.
    """

    def __init__(self, buffer_shape, tile_shape, index_map):
        """
        Initialize a TileView object.

        Parameters
        ----------
        buffer_shape : list of int or PrimExpr
            The shape of the buffer being tiled.
        tile_shape : list of int or PrimExpr
            The shape of each tile.
        index_map : list of int
            Which dimensions of the buffer are tiled.
            Supports negative indices (e.g., -1 for last dim, -2 for second to last).

        Tail tiles are represented explicitly: tiled dimensions use ceildiv,
        and later lowering attaches predicates to the resulting accesses.
        """
        self.__init_handle_by_constructor__(_ffi_api.TileView, buffer_shape, tile_shape, index_map)

    @property
    def tile_shape(self):
        """
        Get the tile shape.

        Returns
        -------
        List[PrimExpr]
            The shape of each tile.
        """
        return _ffi_api.TileView_tile_shape(self)

    @property
    def index_map(self):
        """
        Get the index map (which dimensions are tiled).

        Returns
        -------
        List[PrimExpr]
            The indices of dimensions being tiled.
        """
        return _ffi_api.TileView_index_map(self)

    @property
    def buffer_shape(self):
        """
        Get the original buffer shape.

        Returns
        -------
        List[PrimExpr]
            The shape of the original buffer.
        """
        return _ffi_api.TileView_buffer_shape(self)

    @property
    def tiled_buffer_shape(self):
        """
        Get the tiled buffer shape.

        The tiled shape is computed by replacing each tiled dimension with
        num_tiles (ceildiv(buffer_dim, tile_dim)), then appending the tile
        dimensions.

        For example, buffer_shape=(64, 128), tile_shape=(16, 32), index_map=(-2, -1)
        results in tiled_buffer_shape=(4, 4, 16, 32).

        Returns
        -------
        List[PrimExpr]
            The shape of the tiled buffer.
        """
        return _ffi_api.TileView_tiled_buffer_shape(self)

    @property
    def vector_lanes(self):
        """
        Compute the vector lanes (total elements per tile).

        Returns
        -------
        PrimExpr
            Product of tile_shape dimensions.
        """
        return _ffi_api.TileView_vector_lanes(self)

    def is_equal(self, other: TileView) -> bool:
        """
        Check if this TileView is equal to another.

        Parameters
        ----------
        other : TileView
            The TileView to compare with.

        Returns
        -------
        bool
            True if the TileViews are equal.
        """
        return _ffi_api.TileView_is_equal(self, other)

    def __repr__(self):
        return (
            f"TileView(buffer_shape={list(self.buffer_shape)}, "
            f"tile_shape={list(self.tile_shape)}, "
            f"index_map={list(self.index_map)}) -> "
            f"tiled_shape={list(self.tiled_buffer_shape)}"
        )


def make_tileview(buffer: Buffer | BufferLoad | BufferRegion, tile_shape, index_map):
    """
    Create a TileView from a buffer, tile shape, and index map.

    Parameters
    ----------
    buffer : Buffer | BufferLoad | BufferRegion
        The buffer being tiled. The shape will be extracted automatically.
    tile_shape : list of int or PrimExpr
        The shape of each tile.
    index_map : list of int
        Which dimensions of the buffer are tiled.
        Supports negative indices (e.g., -1 for last dim).

    Returns
    -------
    TileView
        A TileView object.

    Examples
    --------
    >>> # Tile a buffer into (16, 32) tiles on the last two dimensions
    >>> tv = make_tileview(my_buffer, [16, 32], [-2, -1])
    >>> print(tv.tiled_buffer_shape)  # e.g., [4, 4, 16, 32] for a (64, 128) buffer
    """
    _, buffer_shape, _ = _get_buffer_info(buffer)
    return _ffi_api.make_tileview(buffer_shape, tile_shape, index_map)

File: src/tileview/tileview.cc
/*!
 * \file tileview/tileview.cc
 * \brief Implementation of TileView abstraction.
 */

#include "tileview.h"

#include <tvm/ffi/reflection/registry.h>
#include <tvm/runtime/logging.h>
#include <tvm/tir/op.h>

namespace tvm {
namespace tl {

using namespace tir;

TileViewNode::TileViewNode(Array<PrimExpr> buffer_shape,
                           Array<PrimExpr> tile_shape,
                           Array<PrimExpr> index_map) {
  ICHECK_EQ(tile_shape.size(), index_map.size())
      << "tile_shape and index_map must have the same size";

  buffer_shape_ = std::move(buffer_shape);
  tile_shape_ = std::move(tile_shape);
  index_map_ = std::move(index_map);

  int ndim = static_cast<int>(buffer_shape_.size());
  arith::Analyzer analyzer;

  // Build a map from original dim index to tile_dim_index
  std::vector<int> dim_to_tile_idx(ndim, -1);
  for (size_t i = 0; i < tile_shape_.size(); ++i) {
    const auto *idx_int = index_map_[i].as<IntImmNode>();
    ICHECK(idx_int) << "index_map must contain integer constants";
    int idx = static_cast<int>(idx_int->value);
    if (idx < 0) {
      idx = ndim + idx;
    }
    ICHECK(idx >= 0 && idx < ndim)
        << "index_map[" << i << "]=" << index_map_[i]
        << " is out of bounds for buffer with " << ndim << " dimensions";
    dim_to_tile_idx[idx] = static_cast<int>(i);
  }

  // Compute tiled_buffer_shape
  // For each original dimension:
  // - If tiled: replace with num_tiles (ceildiv(buffer_dim, tile_dim))
  // - If not tiled: keep as is
  // Then append all tile dimensions at the end

  Array<PrimExpr> tiled_shape;
  for (int d = 0; d < ndim; ++d) {
    if (dim_to_tile_idx[d] >= 0) {
      int tile_idx = dim_to_tile_idx[d];
      PrimExpr buf_dim = buffer_shape_[d];
      PrimExpr tile_dim = tile_shape_[tile_idx];

      PrimExpr num_tiles = analyzer.Simplify(ceildiv(buf_dim, tile_dim));
      tiled_shape.push_back(num_tiles);
    } else {
      tiled_shape.push_back(buffer_shape_[d]);
    }
  }

  // Append tile shape dimensions at the end
  for (const auto &ts : tile_shape_) {
    tiled_shape.push_back(ts);
  }

  tiled_buffer_shape_ = std::move(tiled_shape);
}

PrimExpr TileViewNode::VectorLanes() const {
  PrimExpr lanes = Integer(1);
  for (const auto &dim : tile_shape_) {
    lanes = lanes * dim;
  }
  return lanes;
}

bool TileViewNode::IsEqual(const TileViewNode *other) const {
  if (other == nullptr)
    return false;
  bool ret = StructuralEqual()(tile_shape_, other->tile_shape_);
  ret &= StructuralEqual()(index_map_, other->index_map_);
  ret &= StructuralEqual()(buffer_shape_, other->buffer_shape_);
  return ret;
}

void TileViewNode::RegisterReflection() {
  namespace refl = tvm::ffi::reflection;
  refl::ObjectDef<TileViewNode>()
      .def_ro("tile_shape", &TileViewNode::tile_shape_)
      .def_ro("index_map", &TileViewNode::index_map_)
      .def_ro("buffer_shape", &TileViewNode::buffer_shape_)
      .def_ro("tiled_buffer_shape", &TileViewNode::tiled_buffer_shape_);
}

TileView::TileView(Array<PrimExpr> buffer_shape, Array<PrimExpr> tile_shape,
                   Array<PrimExpr> index_map) {
  auto n = tvm::ffi::make_object<TileViewNode>(
      std::move(buffer_shape), std::move(tile_shape), std::move(index_map));
  data_ = std::move(n);
}

TileView makeTileView(Array<PrimExpr> buffer_shape, Array<PrimExpr> tile_shape,
                      Array<PrimExpr> index_map) {
  return TileView(std::move(buffer_shape), std::move(tile_shape),
                  std::move(index_map));
}

TVM_FFI_STATIC_INIT_BLOCK() {
  namespace refl = tvm::ffi::reflection;
  refl::GlobalDef()
      .def_packed("tl.TileView",
                  [](PackedArgs args, Any *rv) {
                    *rv = TileView(args[0].cast<Array<PrimExpr>>(),
                                   args[1].cast<Array<PrimExpr>>(),
                                   args[2].cast<Array<PrimExpr>>());
                  })
      .def("tl.TileView_tile_shape",
           [](TileView tv) { return tv->TileShape(); })
      .def("tl.TileView_index_map", [](TileView tv) { return tv->IndexMap(); })
      .def("tl.TileView_buffer_shape",
           [](TileView tv) { return tv->BufferShape(); })
      .def("tl.TileView_tiled_buffer_shape",
           [](TileView tv) { return tv->TiledBufferShape(); })
      .def("tl.TileView_vector_lanes",
           [](TileView tv) { return tv->VectorLanes(); })
      .def("tl.TileView_is_equal",
           [](TileView tv, TileView other) {
             return tv->IsEqual(other.as<TileViewNode>());
           })
      .def("tl.make_tileview",
           [](Array<PrimExpr> buffer_shape, Array<PrimExpr> tile_shape,
              Array<PrimExpr> index_map) {
             return makeTileView(buffer_shape, tile_shape, index_map);
           });
}

TVM_FFI_STATIC_INIT_BLOCK() { TileViewNode::RegisterReflection(); }

} // namespace tl
} // namespace tvm
